modules/runner/src/main/java/org/apache/ignite/internal/app/IgnitionImpl.java

```diff
@@ -164,20 +189,29 @@ private static Ignite doStart(String nodeName, @Nullable String cfgContent, Path
         ackBanner();
 
         try {
-            nodeToStart.start(cfgContent);
+            return nodeToStart.start(cfgContent)
+                    .handle((ignite, e) -> {
+                        if (e == null) {
+                            ackSuccessStart();
+
+                            return ignite;
+                        } else {
+                            throw handleException(nodeName, e);
+                        }
+                    });
         } catch (Exception e) {
-            nodes.remove(nodeName);
-
-            if (e instanceof IgniteException) {
-                throw e;
-            } else {
-                throw new IgniteException(e);
-            }
+            throw handleException(nodeName, e);
         }
+    }
 
-        ackSuccessStart();
+    private static IgniteException handleException(String nodeName, Throwable e) {
```

Comment:
I suggest to rename the method to `handleStartException()` to make it clear that it's about handling start exceptions, not any exception at all (which could be tempting later for someone who is not attentive enough).